internal/cloud/azure/application_identity.go

```diff
@@ -78,13 +79,14 @@ func CreateApplicationIdentity(ctx context.Context, config *ApplicationIdentityC
 	if spErr != nil {
 		return spErr
 	}
-	config.ServicePrincipalID = *sp.ID
-	config.ID = *sp.ID
+	config.ServicePrincipalObjectID = *sp.ID
+	config.ServicePrincipalClientID = *sp.AppId
```

Comment:
Having this available is the actual "fix" for this PR. The rest is trying to make it easier to understand. 

When the app is created or fetched above, here's a snippet of the payload.

```
HTTP/1.1 200 OK
Content-type: application/json

{
    "@odata.context": "https://graph.microsoft.com/v1.0/$metadata#applications/$entity",
    "id": "03ef14b0-ca33-4840-8f4f-d6e91916010e",
    "deletedDateTime": null,
    "isFallbackPublicClient": null,
    "appId": "631a96bc-a705-4eda-9f99-fdaf9f54f6a2",
```

After that, when we create the service principal, that _also_ has an `appId`. It's my understanding that this is true:

+ app.ID == sp.ID
+ app.AppID == sp.AppID

Exposing the client ID (`*app.AppId`) gets us the right creds passed into the app's env-vars. 

```
app_settings = {
     DOCKER_REGISTRY_SERVER_URL      = data.azurerm_container_registry.main.login_server
     DOCKER_REGISTRY_SERVER_USERNAME = module.application.azure_application_identity.service_principal_client_id
     DOCKER_REGISTRY_SERVER_PASSWORD = module.application.azure_application_identity.service_principal_secret
   }
```

Reply:
I think you're right that app.AppId == sp.AppId, but I don't think it's true that app.Id == sp.Id.

Reply:
> I could bind roles to the sp.Id, but not to the app.Id.

We were binding roles using the `sp.Id`, but using the `sp.Id` as the principal user to auth to ACR was failing, because it needed the principal ID of the service principal, which via the API is generated via `sp.AppId`